File: src/ReverseNodesInKGroup.java
/**
 * User: guoxc
 * Date: 15-1-12
 * Time: 22:32
 */
public class ReverseNodesInKGroup {
    public ListNode reverseKGroup(ListNode head, int k){
        if(k == 1) return head;
        ListNode c1 = head;
        ListNode c2 = head;
        ListNode res = head;
        int i = 0;
        while(c1 != null && i < k){
            res = c1;
            c1 = c1.next;
            i += 1;
        }
        if(i == k){
            ListNode t1 = head;
            ListNode t2 = head.next;
            ListNode t3 = null;
            t1.next = null;
            for(int j = 0; j < k-1; j += 1){
                t3 = t2.next;
                t2.next = t1;
                t1 = t2;
                t2 = t3;
            }
            c2.next = reverseKGroup(c1,k);
            return res;
        }else{
            return head;
        }
    }
    public static void test(){
        ListNode l = new ListNode(new int[]{1,2,3,4,5});
        ListNode l1 = new ListNode(new int[]{1});
        ListNode l2 = new ListNode(new int[]{1,2});
        System.out.println(new ReverseNodesInKGroup().reverseKGroup(l2,2));
    }
}

File: cpp/reverse_nodes_in_k_group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
private:
    ListNode* reverse(ListNode* src){
        ListNode* p1 = src;
        ListNode* p2 = src->next;
        if(p2 == NULL) return src;
        ListNode* res = p2;
        ListNode* p3;
        p1->next = NULL;
        while(p1 != NULL && p2 != NULL){
            p3 = p2->next;
            p2->next = p1;
            p1 = p2;
            res = p1;
            p2 = p3;
        }
        return res;;
    }
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        int count = 0;
        if(k <= 1) return head;
        ListNode* p = head;
        ListNode* last_head = head;
        ListNode* res = NULL;
        ListNode* res_tail = NULL;
        while(p != NULL){
            if(count == k - 1){
                ListNode* p_next = p->next; //save for keepgoing
                p->next = NULL; // break;
                ListNode* new_list_head = reverse(last_head);
                if(res == NULL){
                    res = new_list_head;
                    res_tail = last_head;
                }else{
                    res_tail->next = new_list_head;
                    res_tail = last_head;
                }
                last_head = p_next;
                p = p_next;
                count = 0;
            }else{
                p = p->next;
                count += 1;
            }
        }
        if(res == NULL) res = last_head;
        if(res_tail != NULL) res_tail->next = last_head;
        return res;
    }
};
